### ISSUE TITLE
Feature Request or Bug? Add support for topology aware resume training?

### ISSUE THREAD

**Issue by tjoymeed** [volcengine/verl#1377]
Hi there,  thanks a lot for the wonderful work!

It seems that the current resume-training-from-checkpoint implementation doesn't take into account the change of training topology. 

For example, the previous run is on 4x8 GPU cards and the next resumed run is on 8x8 GPU cards.

The current implementation will raise an error.

But it seems that PyTorch fsdp supports the following:

Topology Agnostic Loading: Handles rank count changes through Stateful object tracking

Tensor Remapping: Automatically redistributes shards using LoadPlanner extensions

Partial Loading: Supports selective parameter loading when architecture changes

Is it possible to add this feature in?

Thanks a lot!

-----------------------

# Core resharding workflow
from torch.distributed.checkpoint import load_state_dict

def load_checkpoint(new_model, checkpoint_path):
    new_state_dict = get_state_dict(new_model)
    load_state_dict(
        state_dict=new_state_dict,
        checkpoint_id=checkpoint_path,
        planner=resharding_planner
    )
    set_state_dict(new_model, new_state_dict)

model = FSDP(
    model,
    sharding_strategy=ShardingStrategy.FULL_SHARD,
    device_id=torch.cuda.current_device(),
    auto_wrap_policy=custom_auto_wrap_policy
)

from torch.distributed.checkpoint import (
    load_state_dict,
    FileSystemReader
)

# Initialize empty state dict matching new topology
model_state_dict, optimizer_state_dict = get_state_dict(model, optimizer)
checkpoint = {"model": model_state_dict, "optimizer": optimizer_state_dict}

# Perform topology-aware loading
load_state_dict(
    state_dict=checkpoint,
    storage_reader=FileSystemReader(checkpoint_dir),
    planner=DefaultLoadPlanner(),
    no_dist=False
)

# Apply loaded state
set_state_dict(model, optim, 
              model_state_dict=checkpoint["model"],
              optim_state_dict=checkpoint["optimizer"])

**Comment by 0x404** [comment]
Yes, you are right, the current checkpoint requires the checkpoint files be the same running topology.

I think we can impl this by having rank 0 load and merge all shards at CPU, then broadcast the merged state to all ranks, this will be a bit slow but should work correctly. I can impl this is the following days.

**Comment by tjoymeed** [comment]
Thank you!
I saw some people are talking about this: https://github.com/ByteDance-Seed/ByteCheckpoint
Maybe we could directly use this product to avoid reinvent the wheels?

**Comment by 0x404** [comment]
Great! I will take a look :)

**Comment by Kirrito-k423** [comment]
### Scheduled issue cleanup / 定时 issue 清理

中文：该 issue 命中定时清理规则：最近一次非当前运行账号的公开回复或引用已经超过 6 个月。因此按定时清理标准先关闭。若作者或社区成员仍能提供新的复现、证据或需求背景，欢迎继续评论请求重开，或创建带有最新信息的新 issue。

证据：
- Issue: https://github.com/verl-project/verl/issues/1377
- 最近一次非当前运行账号的公开回复或公开引用是 `2025-05-05T07:52:29Z` UTC（comment by @0x404），截至 2026-07-05 已超过 6 个月。
- 已检查 comments 来源：https://api.github.com/repos/verl-project/verl/issues/1377/comments
- 已检查 timeline 来源：https://api.github.com/repos/verl-project/verl/issues/1377/timeline

English: This issue matches the scheduled cleanup rule: the latest public reply or reference by an account other than the current runner is older than 6 months. Therefore, this is being closed under the scheduled cleanup policy. If the author or community members can still provide a fresh reproduction, evidence, or requirement context, please comment to request reopening or open a new issue with updated information.

Evidence:
- Issue: https://github.com/verl-project/verl/issues/1377
- The latest public reply or public reference from an account other than the current runner was `2025-05-05T07:52:29Z` UTC (comment by @0x404), which is more than 6 months before 2026-07-05.
- Checked comments source: https://api.github.com/repos/verl-project/verl/issues/1377/comments
- Checked timeline source: https://api.github.com/repos/verl-project/verl/issues/1377/timeline

This response was generated by a scheduled Codex GPT5.5 issue processing task.
